spec/lang/step/operators.md

```diff
@@ -346,28 +346,67 @@ impl<M: Memory> Machine<M> {
         };
         (Value::Bool(b), Type::Bool)
     }
-    /// Prepare a value for being compared with a `RelOp`. This involves turning it into a pair of `Int`s.
-    /// This pair is compared lexicographically, and is only important for wide pointers.
-    fn prepare_for_rel_op(val: Value<M>) -> (Int, Int) {
+    /// Prepare a value for being compared with a `RelOp`. This involves turning it into an `Int`.
+    fn prepare_for_int_rel_op(val: Value<M>) -> Int {
         match val {
-            Value::Int(i) => (i, Int::ZERO),
-            Value::Bool(b) => (if b { Int::from(1) } else { Int::from(0) }, Int::ZERO),
-            Value::Ptr(Pointer { thin_pointer, metadata: None}) => (thin_pointer.addr, Int::ZERO),
-            Value::Ptr(Pointer { thin_pointer, metadata: Some(PointerMeta::ElementCount(i))}) => (thin_pointer.addr, i),
-            Value::Ptr(Pointer { thin_pointer, metadata: Some(PointerMeta::VTablePointer(v))}) => (thin_pointer.addr, v.addr),
-            _ => panic!("invalid value for relational operator"),
+            Value::Int(i) => i,
+            Value::Bool(b) => if b { Int::from(1) } else { Int::from(0) },
+            _ => panic!("invalid value for integer relational operator"),
         }
     }
+
+    fn eval_ptr_rel_op(rel: RelOp, left: Pointer<M::Provenance>, right: Pointer<M::Provenance>) -> (Value<M>, Type) {
+        use RelOp::*;
+        let cmp = Self::compare_ptr(left, right);
+        match rel {
+            Lt => (Value::Bool(cmp.is_lt()), Type::Bool),
+            Gt => (Value::Bool(cmp.is_gt()), Type::Bool),
+            Le => (Value::Bool(cmp.is_le()), Type::Bool),
+            Ge => (Value::Bool(cmp.is_ge()), Type::Bool),
+            Eq => (Value::Bool(cmp.is_eq()), Type::Bool),
+            Ne => (Value::Bool(cmp.is_ne()), Type::Bool),
```

Comment:
I like the idea of using `std::cmp::Ordering`; it should be possible to use that to avoid having to match on `rel` twice. Basically, in `eval_bin_op`, we do something like
```rust
let ord = match l_ty { ... };
match rel {
   // Convert `ord` to ` Value::Bool`/`Value::Int` according to `rel`.
}
```
Then `rel` should not be used anywhere else.